File: algorithms/algorithms-cplusplus/leetcode/Question_0728_Self_Dividing_Numbers.cc
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> selfDividingNumbers(int left, int right) {
    vector<int> res;
    for (int i = left; i <= right; i++) {
      if (isSelfDividing(i)) {
        res.push_back(i);
      }
    }
    return res;
  }

 private:
  bool isSelfDividing(int n) {
    string s = to_string(n);
    for (char c : s) {
      if (c == '0' || n % (c - '0') != 0) {
        return false;
      }
    }
    return true;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0728_Self_Dividing_Numbers.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Question_0728_Self_Dividing_Numbers {
    public List<Integer> selfDividingNumbers(int left, int right) {
        List<Integer> res = new ArrayList();
        for (int n = left; n <= right; ++n) {
            if (selfDividing(n)) {
                res.add(n);
            }
        }
        return res;
    }

    public boolean selfDividing(int n) {
        for (char c : String.valueOf(n).toCharArray()) {
            if (c == '0' || (n % (c - '0') > 0)) {
                return false;
            }
        }
        return true;
    }
}
